File: problems/Integer to English Words/numberToWords.cs
public class Solution {
    public string NumberToWords(int num) {
        if (0 == num) {
            return "Zero";
        }

        List<string> result = new List<string>();
        int billions = num / 1000000000;
        num -= billions * 1000000000;
        int millions = num / 1000000;
        num -= millions * 1000000;
        int thousands = num / 1000;
        num -= thousands * 1000;

        if (0 < billions) {
            result.Add(stringFromTreeDigits(billions));
            result.Add("Billion");
        }

        if (0 < millions) {
            result.Add(stringFromTreeDigits(millions));
            result.Add("Million");
        }

        if (0 < thousands) {
            result.Add(stringFromTreeDigits(thousands));
            result.Add("Thousand");
        }

        if (0 < num) {
            result.Add(stringFromTreeDigits(num));
        }

        return string.Join(" ", result);
    }

    private string stringFromTreeDigits(int num) {
        List<string> result = new List<string>();
        int hundreds = num / 100;

        num -= hundreds * 100;

        if (0 < hundreds) {
            result.Add(getDigit(hundreds));
            result.Add("Hundred");
        }

        if (0 < num) {
            if (20 <= num) {
                result.Add(getTens(num / 10));
                if (0 != num % 10) {
                    result.Add(getDigit(num % 10));
                }
            } else if (10 <= num && 20 > num) {
                result.Add(getExtras(num));
            } else {
                result.Add(getDigit(num));
            }
        }

        return string.Join(" ", result);
    }

    private string getDigit(int num) {
        var digits = new Dictionary<int, string>{{1, "One"}, {2, "Two"}, {3, "Three"}, {4, "Four"}, {5, "Five"}, {6, "Six"}, {7, "Seven"}, {8, "Eight"}, {9, "Nine"}};

        return digits[num];
    }

    private string getExtras(int num) {
        var extras = new Dictionary<int, string>{{10,"Ten"},{11,"Eleven"},{12,"Twelve"},{13,"Thirteen"},{14,"Fourteen"},{15,"Fifteen"},{16,"Sixteen"},{17,"Seventeen"},{18,"Eighteen"},{19,"Nineteen"}};

        return extras[num];
    }

    private string getTens(int num) {
        var tens = new Dictionary<int, string>{{2,"Twenty"},{3,"Thirty"},{4,"Forty"},{5,"Fifty"},{6,"Sixty"},{7,"Seventy"},{8,"Eighty"},{9,"Ninety"}};

        return tens[num];
    }
}

File: problems/Integer to English Words/numberToWords.java
class Solution {
    public String numberToWords(int num) {
        if (0 == num) {
            return "Zero";
        }

        List<String> result = new ArrayList<>();
        int billions = num / 1000000000;
        num -= billions * 1000000000;
        int millions = num / 1000000;
        num -= millions * 1000000;
        int thousands = num / 1000;
        num -= thousands * 1000;

        if (0 < billions) {
            result.add(getStringFromThreeDigits(billions));
            result.add("Billion");
        }

        if (0 < millions) {
            result.add(getStringFromThreeDigits(millions));
            result.add("Million");
        }

        if (0 < thousands) {
            result.add(getStringFromThreeDigits(thousands));
            result.add("Thousand");
        }

        if (0 < num) {
            result.add(getStringFromThreeDigits(num));
        }

        return String.join(" ", result);
    }

    private String getStringFromThreeDigits(int num) {
        List<String> result = new ArrayList<>();
        int hundreds = num / 100;

        num -= hundreds * 100;

        if (0 < hundreds) {
            result.add(getStringFromOneDigit(hundreds));
            result.add("Hundred");
        }

        if (0 < num) {
            if (20 <= num) {
                int tens = num / 10;

                num -= tens * 10;
                result.add(getTens(tens));

                if (0 < num) {
                    result.add(getStringFromOneDigit(num));
                }
            } else if (10 <= num && 20 > num) {
                result.add(getExtras(num));
            } else {
                result.add(getStringFromOneDigit(num));
            }
        }

        return String.join(" ", result);
    }

    private String getStringFromOneDigit(int num) {
        switch (num) {
            case 1:
                return "One";
            case 2:
                return "Two";
            case 3:
                return "Three";
            case 4:
                return "Four";
            case 5:
                return "Five";
            case 6:
                return "Six";
            case 7:
                return "Seven";
            case 8:
                return "Eight";
            case 9:
                return "Nine";
            default:
                return "Error";
        }
    }

    private String getTens(int num) {
        switch (num) {
            case 2:
                return "Twenty";
            case 3:
                return "Thirty";
            case 4:
                return "Forty";
            case 5:
                return "Fifty";
            case 6:
                return "Sixty";
            case 7:
                return "Seventy";
            case 8:
                return "Eighty";
            case 9:
                return "Ninety";
            default:
                return "Error";
        }
    }

    private String getExtras(int num) {
        switch (num) {
            case 10:
                return "Ten";
            case 11:
                return "Eleven";
            case 12:
                return "Twelve";
            case 13:
                return "Thirteen";
            case 14:
                return "Fourteen";
            case 15:
                return "Fifteen";
            case 16:
                return "Sixteen";
            case 17:
                return "Seventeen";
            case 18:
                return "Eighteen";
            case 19:
                return "Nineteen";
            default:
                return "Error";
        }
    }
}
